File: netforce_general/netforce_general/controllers/export.py
# Copyright (c) 2012-2015 Netforce Co. Ltd.
# 
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
# 
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
# IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
# DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
# OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE
# OR OTHER DEALINGS IN THE SOFTWARE.

from netforce.controller import Controller
from netforce.model import get_model,clear_cache,fields
from netforce import database
from netforce import template
from netforce.action import get_action
from netforce.utils import get_data_path,set_data_path
from netforce.database import get_connection,get_active_db
from netforce import config
from netforce import static
from netforce import access
import json
from pprint import pprint
import os
import base64
import urllib
import netforce
import sys
import tempfile
from lxml import etree
import time

def parse_args(handler):
    res={}
    for path in handler.request.arguments:
        for v in handler.get_arguments(path):
            if v=="":
                continue
            set_data_path(res,path,v)
    return res

class Export(Controller): # TODO: cleanup
    _path="/export"

    def get(self):
        db=get_connection()
        if db:
            db.begin()
        try:
            clear_cache()
            ctx={
                "request": self.request,
                "request_handler": self,
                "dbname": get_active_db(),
            }
            data=self.get_cookies()
            if data:
                ctx.update(data)
            action_vals=parse_args(self)
            ctx.update(action_vals)
            name=action_vals.get("name")
            if name:
                action_ctx=action_vals
                action=get_action(name,action_ctx)
                for k,v in action.items():
                    if k not in action_vals:
                        action_vals[k]=v
            if "context" in action_vals:
                ctx.update(action_vals["context"])
            action_vals["context"]=ctx
            self.clear_flash()
            type=action_vals.get("type","view")
            if type=="export":
                print("XXX export")
                model=action_vals["model"]
                m=get_model(model)
                ids=action_vals.get("ids")
                if ids:
                    if ids[0]=="[": # XXX
                        ids=ids[1:-1]
                    ids=[int(x) for x in ids.split(",")]
                else:
                    condition=action_vals.get("condition")
                    if condition:
                        print("condition",condition)
                        condition=json.loads(condition)
                        ids=m.search(condition)
                    else:
                        ids=m.search([]) # XXX
                ctx=action_vals.copy()
                if ctx.get("export_fields"):
                    if isinstance(ctx["export_fields"],str):
                        ctx["export_fields"]=json.loads(ctx["export_fields"])
                else:
                    try:
                        view=get_xml_view(model=model,type="export")
                        doc=etree.fromstring(view["layout"])
                        field_names=[]
                        for el in doc.iterfind(".//field"):
                            name=el.attrib["name"]
                            field_names.append(name)
                        ctx["export_fields"]=field_names
                    except: # default export fields
                        req_field_names=[]
                        other_field_names=[]
                        for n,f in m._fields.items():
                            if isinstance(f,(fields.One2Many,fields.Many2Many)):
                                continue
                            if isinstance(f,fields.Json):
                                continue
                            if not f.store and not f.function:
                                continue
                            if f.required:
                                req_field_names.append(n)
                            else:
                                other_field_names.append(n)
                        ctx["export_fields"]=sorted(req_field_names)+sorted(other_field_names)
                data=m.export_data(ids,context=ctx)
                db=get_connection()
                if db:
                    db.commit()
                filename=action_vals.get("filename","export.csv")
                self.set_header("Content-Disposition","attachment; filename=%s"%filename)
                self.set_header("Content-Type","text/csv")
                self.write(data)
            else:
                raise Exception("Invalid action type: %s"%type)
        except Exception as e:
            import traceback
            traceback.print_exc(file=sys.stdout)
            db=get_connection()
            if db:
                db.rollback()
            raise e

Export.register()

File: netforce_ui/netforce_ui/views/export.js
/*
 * Copyright (c) 2012-2015 Netforce Co. Ltd.
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
 * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
 * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE
 * OR OTHER DEALINGS IN THE SOFTWARE.
 */

var Export=NFView.extend({
    _name: "export",
    events: {
        "click .field-row": "click_field",
        "click .select-all": "click_select_all",
        "click .do-export": "do_export"
    },

    render: function() {
        log("export.render");
        var that=this;
        var model_cls=get_model(this.options.model);
        this.data.model_name=this.options.model;
        this.data.model_string=model_cls.string;
        this.data.fields=[{
            name: "id",
            string: "Database ID",
            row_id: _.uniqueId("row")
        }];
        _.each(model_cls.fields,function(f,n) {
            var field={
                name: n,
                string: f.string,
                row_id: _.uniqueId("row")
            };
            if (f.type=="many2one") {
                field.can_expand=true;
            } else if (f.type=="one2many") {
                field.can_expand=true;
            }
            that.data.fields.push(field);
        });
        this.data.fields=_.sortBy(this.data.fields,function(f) {return f.string});
        NFView.prototype.render.call(this);
        //this.load_export_view();
        this.load_export_fields();
    },

    click_field: function(e) {
        log("export.click_field");
        log("target",e.target);
        if ($(e.target).prop("tagName")=="INPUT") return;
        e.preventDefault();
        var tr=$(e.target).parents("tr");
        var state=tr.data("state");
        if (state=="collapsed") {
            this.expand_field(tr);
        } else if (state=="expanded") {
            this.collapse_field(tr);
        }
    },

    expand_field: function(tr) {
        var model=tr.data("model");
        var field=tr.data("field");
        var depth=tr.data("depth");
        var path=tr.data("path");
        log("model",model,"field",field);
        var model_cls=get_model(model);
        var f=model_cls.fields[field];
        var rmodel=f.relation;
        var rmodel_cls=get_model_cls(rmodel);
        var rdepth=depth+1;
        var fields_=[];
        _.each(rmodel_cls.fields,function(f,n) {
            var field={
                name: n,
                string: f.string,
                row_id: _.uniqueId("row")
            };
            if (f.type=="many2one") {
                field.can_expand=true;
            } else if (f.type=="one2many") {
                field.can_expand=true;
            }
            fields_.push(field);
        });
        fields_=_.sortBy(fields_,function(f) {return f.string});
        var last_el=tr;
        _.each(fields_,function(f) {
            var html="<tr class='field-row' data-field='"+f.name+"' data-model='"+rmodel+"' data-depth='"+rdepth+"' data-row-id='"+f.row_id+"' data-state='collapsed' data-parent-id='"+tr.data("row-id")+"' data-path='"+path+"."+f.name+"'><td><input type='checkbox'/></td><td style='padding-left:"+rdepth*30+"px'>";
            if (f.can_expand) {
                html+="<i class='icon-chevron-right' style='margin-right:5px'></i>";
            }
            html+=f.string+"</td><td>"+f.name+"</td><td>"+rmodel_cls.string+"</td></tr>";
            var el=$(html);
            last_el.after(el);
            last_el=el;
        });
        tr.data("state","expanded");
    },

    collapse_field: function(tr) {
        var row_id=tr.data("row-id");
        this.$el.find("tr[data-parent-id="+row_id+"]").remove();
        tr.data("state","collapsed");
    },

    click_select_all: function(e) {
        log("export.click_select_all",e.target);
        var res=$("tr.field-row input:not(:checked)");
        if (res.length>0) {
            res.prop("checked",true);
            $(e.target).prop("checked",true);
        } else {
            $("tr.field-row input").prop("checked",false);
            $(e.target).prop("checked",false);
        }
    },

    do_export: function(e) {
        log("export.do_export");
        e.preventDefault();
        var export_fields=[];
        $("tr.field-row input:checked").each(function() {
            var path=$(this).parents("tr").data("path");
            export_fields.push(path);
        });
        log("export_fields",export_fields);
        this.export_fields=export_fields;
        this.save_export_fields();
        var action={
            name: "export_data", // XXX: should not need name
            type: "export",
            model: this.options.model,
            condition: this.options.condition,
            export_fields: export_fields
        };
        exec_action(action);
    },

    select_fields: function(fields) {
        log("export.select_fields",fields);
        var that=this;
        _.each(fields,function(name) {
            log("name",name);
            var comps=name.split(".");
            for (var i=0; i<comps.length-1; i++) {
                var path=comps.splice(0,i+1).join(".");
                log("parent path",path);
                var tr=that.$el.find("tr.field-row[data-path='"+path+"']")
                var state=tr.data("state");
                if (state=="collapsed") {
                    that.expand_field(tr);
                }
            }
            var inp=that.$el.find("tr.field-row[data-path='"+name+"']").find("input");
            log("inp",inp);
            inp.prop("checked",true);
        });
    },

    load_export_view: function() {
        var export_view=get_xml_layout({model:this.options.model,type:"export",noerr:true});
        if (!export_view) return;
        var layout=export_view.layout;
        var doc=$.parseXML(layout);
        var $layout=$(doc).children();
        var fields=[];
        $layout.find("field").each(function() {
            var $el=$(this);
            var name=$el.attr("name");
            fields.push(name);
        });
        this.select_fields(fields);
    },

    save_export_fields: function() {
        rpc_execute("field.default","set_default",[this.options.model,"_export_fields",JSON.stringify(this.export_fields)]);
    },

    load_export_fields: function() {
        var that=this;
        rpc_execute("field.default","get_default",[this.options.model,"_export_fields"],{},function(err,data) {
            if (!data) return;
            var fields=JSON.parse(data);
            that.select_fields(fields);
        });
    }
});

Export.register();
